feat: A2A task tracking and query observability improvements

gabrieleblue opened the pull request:
## Summary
- Add A2ATask CRD for tracking agent-to-agent protocol task execution
- Merge main branch updates including query conditions and RAG samples
- Handle blocking A2A tasks by extracting response text AND creating tracking resources
- Add both `associatedTask` and `conditions` fields to Query status for comprehensive observability

## Key Changes
- **A2A Task Tracking**: New A2ATask resource tracks task lifecycle, history, and completion state
- **Blocking Task Support**: Extract final response from completed/failed tasks while creating tracking resources
- **Query Observability**: Support both A2A task references and standard K8s conditions
- **Flexible Input**: Support preserve-unknown-fields for complex message types

Commit:
udpate

Commit:
updating a2a exexution

Commit:
update

dwmkerr committed:
fix: merge main into feat/a2a-task-crd

Resolved conflicts by combining features from both branches:

- **A2A task handling**: For blocking tasks, extract text response AND create A2ATask resource for tracking
- **Query CRD**: Keep both `associatedTask` field (A2A tracking) and `conditions` array (standard K8s observability)
- **Input field**: Use `x-kubernetes-preserve-unknown-fields` for flexible message type support

The combined approach provides:
1. Immediate response text for blocking A2A tasks
2. K8s resource tracking via A2ATask for all task-based responses
3. Standard K8s conditions for query observability

dwmkerr committed:
fix: add missing Recorder to ModelReconciler

The ModelReconciler was calling r.Recorder.Event() but the Recorder
was never initialized in main.go, causing nil pointer panics when
models were probed.

dwmkerr committed:
docs: add A2A queries documentation and message context test

dwmkerr committed:
Merge branch 'main' of github.com:mckinsey/agents-at-scale-ark into feat/a2a-docs-and-tests

dwmkerr committed:
debug: add logging to investigate missing phase field in query responses

dwmkerr committed:
fix: add missing phase and raw fields to Query Response CRD schema

The dist CRD was missing phase and raw fields in the Response schema,
causing these fields to be dropped when persisting query responses.
This prevented error messages from appearing in query condition messages.

- Add phase field to track response status (error/done)
- Add raw field for structured response data
- Add description field for API documentation
- Preserve associatedTask field for A2A functionality

dwmkerr committed:
fix: sync dist Query CRD with main, preserving A2A associatedTask field

The dist CRD was out of sync with main, missing critical fields:

Missing fields restored from main:
- spec.type field (enum: [user, messages], default: user)
  This was causing 'cannot unmarshal array into Go value' errors when
  users tried to use type: messages with array input
- spec.type printer column for kubectl output
- parameters.queryParameterRef for Query parameter references

Evaluation-related fields removed (not yet in main):
- spec.evaluatorSelector
- spec.evaluators
- status.evaluations
- status.phase 'evaluating' enum value
- evaluations count printer column

A2A-specific field preserved:
- status.associatedTask (tracks A2ATask for this query)

This brings the dist CRD in sync with main while preserving the
A2A functionality needed for the feature branch.

dwmkerr committed:
fix: implement A2A contextID annotation and fix response handling

- Add ark.mckinsey.com/a2a-context-id annotation constant
- Extract contextID from A2A protocol Message and Task responses
- Store contextID in TokenUsageCollector for upward propagation
- Update Query annotations with A2A contextID after execution
- Simplify A2A function hierarchy to single ExecuteA2AAgent function
- Fix metadata update ordering to prevent status field overwrites
- Remove unnecessary info logs for expected A2A operations
- Add nil checks for optional tokenCollector parameter

dwmkerr committed:
chore: merge main into feat/a2a-task-crd to sync branches

dwmkerr committed:
chore: merge updated feat/a2a-task-crd into feat/a2a-docs-and-tests

dwmkerr commented on ark/api/v1alpha1/a2atask_types.go:
```diff
@@ -0,0 +1,319 @@
+	"trpc.group/trpc-go/trpc-a2a-go/protocol"
+)
+
+type AgentRef struct {
```
Do we have this elsewhere

dwmkerr commented on ark/api/v1alpha1/a2atask_types.go:
```diff
@@ -0,0 +1,319 @@
+}
+
+// A2ATaskPart represents content parts compatible with A2A protocol
+type A2ATaskPart struct {
```
all fields need clear doc strings

dwmkerr commented on ark/api/v1alpha1/a2atask_types.go:
```diff
@@ -0,0 +1,319 @@
+}
+
+// A2ATaskStatus represents task status from A2A protocol
+type A2ATaskTaskStatus struct {
```
let's thnk about how to avoid `TaskTask`

dwmkerr commented on ark/api/v1alpha1/a2atask_types.go:
```diff
@@ -0,0 +1,319 @@
+	Priority int32 `json:"priority,omitempty"`
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default="5m"
+	Timeout *metav1.Duration `json:"timeout,omitempty"`
```
do we need a TTL consistent with query, is this a timeout for the entire time to track the task? if so then 12 hours is likely better.

Do we have a link to the a2a server?

dwmkerr commented on ark/api/v1alpha1/a2atask_types.go:
```diff
@@ -0,0 +1,319 @@
+	// +kubebuilder:validation:Optional
+	CompletionTime *metav1.Time `json:"completionTime,omitempty"`
+	// +kubebuilder:validation:Optional
+	Error string `json:"error,omitempty"`
```
We need conditions, similar to query, suggest which ones

dwmkerr commented on ark/api/v1alpha1/a2atask_types.go:
```diff
@@ -0,0 +1,319 @@
+
+// ConvertPartFromProtocol converts a protocol.Part to A2ATaskPart
+func ConvertPartFromProtocol(part interface{}) A2ATaskPart {
+	switch p := part.(type) {
```
let's move this into an a2a_protocol module

dwmkerr commented on ark/api/v1alpha1/query_types.go:
```diff
@@ -108,6 +108,12 @@ type TokenUsage struct {
 	TotalTokens      int64 `json:"totalTokens,omitempty"`
 }
 
+// A2ATaskReference represents a reference to an associated A2ATask
```
we will remove the associated task for now

dwmkerr commented on ark/api/v1alpha1/query_types.go:
```diff
@@ -119,6 +125,9 @@ type QueryStatus struct {
 	TokenUsage TokenUsage         `json:"tokenUsage,omitempty"`
 	// +kubebuilder:validation:Optional
 	Duration *metav1.Duration `json:"duration,omitempty"`
+	// +kubebuilder:validation:Optional
```
remove

dwmkerr commented on ark/cmd/main.go:
```diff
@@ -225,17 +225,18 @@ func setupControllers(mgr ctrl.Manager) {
 		reconciler interface{ SetupWithManager(ctrl.Manager) error }
 	}{
-		{"Agent", &controller.AgentReconciler{Client: mgr.GetClient(), Scheme: mgr.GetScheme(), Recorder: mgr.GetEventRecorderFor("agent-controller")}},
+		{"Agent", &controller.AgentReconciler{Client: mgr.GetClient(), Scheme: mgr.GetScheme()}},
```
merge error - bug; we need the event recorder

dwmkerr commented on ark/cmd/main.go:
```diff
@@ -225,17 +225,18 @@ func setupControllers(mgr ctrl.Manager) {
 		{"A2AServer", &controller.A2AServerReconciler{Client: mgr.GetClient(), Scheme: mgr.GetScheme(), Recorder: mgr.GetEventRecorderFor("a2aserver-controller")}},
 		{"MCPServer", &controller.MCPServerReconciler{Client: mgr.GetClient(), Scheme: mgr.GetScheme(), Recorder: mgr.GetEventRecorderFor("mcpserver-controller")}},
-		{"Model", &controller.ModelReconciler{Client: mgr.GetClient(), Scheme: mgr.GetScheme(), Recorder: mgr.GetEventRecorderFor("model-controller")}},
+		{"Model", &controller.ModelReconciler{Client: mgr.GetClient(), Scheme: mgr.GetScheme()}},
```
merge error we need recorder

dwmkerr commented on ark/config/crd/bases/ark.mckinsey.com_queries.yaml:
```diff
@@ -57,6 +57,12 @@ spec:
                 description: Input can be a string (type=user) or []openai.ChatCompletionMessageParamUnion
                   (type=messages)
                 x-kubernetes-preserve-unknown-fields: true
+              type:
```
why is this missing in main

dwmkerr commented on ark/config/crd/bases/ark.mckinsey.com_queries.yaml:
```diff
@@ -255,6 +261,15 @@ spec:
             type: object
           status:
             properties:
+              associatedTask:
```
no needed

dwmkerr commented on ark/config/rbac/a2atask_rbac.yaml:
```diff
@@ -0,0 +1,55 @@
+---
+# Allow default service accounts in user namespaces to create A2ATask resources
```
do we need this in ark-tenant

dwmkerr commented on ark/config/rbac/kustomization.yaml:
```diff
@@ -44,4 +45,7 @@ resources:
 - agent_admin_role.yaml
 - agent_editor_role.yaml
 - agent_viewer_role.yaml
+# Tenant namespace permissions for quickstart scenarios (applied separately due to namespace: default)
```
what is this?

dwmkerr commented on ark/dist/chart/templates/crd/ark.mckinsey.com_queries.yaml:
```diff
@@ -21,15 +21,15 @@ spec:
   scope: Namespaced
   versions:
   - additionalPrinterColumns:
```
merge error, we need type, we remove evals, we remove evalselector, we meed query params, this query spec should be same as main

dwmkerr commented on ark/internal/genai/context_utils.go:
```diff
@@ -46,6 +46,15 @@ func getSessionID(ctx context.Context) string {
 	return ""
 }
 
+func getQueryName(ctx context.Context) string {
```
do we need this

dwmkerr commented on docs/content/reference/crds.mdx:
```diff
@@ -5,40 +5,188 @@ description: Detailed specifications for ARK's custom resources
 # Custom Resource Definitions (CRDs)
 
-This page provides detailed specifications for each ARK custom resource.
+This page provides detailed specifications for each ARK custom resource. For an overview of how these resources work with Kubernetes, see [Kubernetes Integration](/concepts/kubernetes).
```
bug, this should be the same as main

av3n93rz committed:
feat: sets HTTP 408 status on stream timeout response

Ensures proper indication of request timeout by sending HTTP 408 status when no stream chunks are received within the expected period.

Improves client-side error handling and clarity for timeout scenarios.

av3n93rz committed:
feat: improves error handling for streaming response failures

Enhances the proxy logic to extract and forward detailed error messages from failed streaming responses, ensuring clients receive more informative SSE error events. Helps with debugging and improves user feedback in case of upstream errors.

av3n93rz committed:
feat: improves error streaming and handling in chat workflows

Enables structured error chunks to be streamed alongside metadata, allowing clients to handle agent errors more gracefully. Updates backend and frontend to parse and relay error details in streaming endpoints for clearer feedback and better API reliability.

av3n93rz committed:
refactor: extract metadata building logic into separate function

dwmkerr committed:
fix: address all 17 PR review comments for A2ATask implementation

- Move protocol conversion functions to internal/genai/a2a_protocol.go
- Remove AssociatedTask field from Query CRD
- Add AgentReconciler event recorder
- Remove ClusterRole-based a2atask_rbac.yaml
- Add A2ATask permissions to ark-tenant chart
- Remove obsolete commented RBAC references
- Fix docs capitalization to match main branch
- Confirm getQueryName function is needed for QueryRef linkage

dwmkerr commented on ark/api/v1alpha1/a2atask_types.go:
```diff
@@ -0,0 +1,230 @@
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=100
+	Progress int32 `json:"progress,omitempty"`
```
is this actually updated used

dwmkerr committed:
fix: remove non-spec Progress field and noisy logs from A2ATask controller

- Remove Progress field from A2ATaskStatus (not part of official A2A spec)
- Remove noisy info logs that fired on every reconciliation
- Clean up unused imports and function parameters

dwmkerr committed:
refactor: consolidate A2A protocol conversion and improve task controller

- Add messageId field to A2ATaskMessage to preserve protocol identifiers
- Move protocol conversion logic to a2a_protocol.go with phase constants
- Consolidate duplicate convertA2AStateToPhase functions into one
- Move MergeArtifacts, MergeHistory, UpdateA2ATaskStatus to protocol layer
- Fix header handling in A2A client creation (previously ignored)
- Rename pollA2ATaskStatus to fetchA2ATaskStatus for clarity
- Remove unused imports and functions from controller
- Use protocol's "canceled" spelling, convert to Ark's "cancelled" phase
- Update both config and dist CRDs with messageId field

dwmkerr commented on ark/internal/controller/a2atask_controller.go:
```diff
@@ -0,0 +1,226 @@
+	}
+
+	// Set start time for new tasks
+	if a2aTask.Status.Phase == genai.PhasePending && a2aTask.Status.StartTime == nil {
```
checking that this is correct

dwmkerr commented on ark/internal/controller/a2atask_controller.go:
```diff
@@ -0,0 +1,226 @@
+	// Fetch task status from A2A server if we have the required information
+	if a2aTask.Status.Phase == genai.PhaseAssigned || a2aTask.Status.Phase == genai.PhaseRunning {
+		if err := r.fetchA2ATaskStatus(ctx, &a2aTask); err != nil {
+			log.Error(err, "failed to fetch A2A task status", "taskId", a2aTask.Spec.TaskID)
```
not needed

dwmkerr committed:
fix: use A2A protocol timestamps for task start and completion times

StartTime and CompletionTime now reflect the actual times from the A2A
server rather than when the Ark controller processed the task. This
provides accurate timing information about when the A2A agent started
and completed execution.

- Parse task.Status.Timestamp from A2A protocol (RFC3339 format)
- Set StartTime when transitioning from pending to running
- Set CompletionTime when reaching terminal state
- Move timestamp parsing to a2a_protocol.go for centralized handling
- Simplify controller to only handle conditions and events

dwmkerr committed:
docs: update A2ATask documentation to reflect implementation changes

- Remove deprecated progress field from examples
- Add messageId field to message history examples
- Clarify field sources (Ark vs A2A protocol) with inline annotations
- Document that timestamps come from A2A protocol
- Explain A2ATask creation for message-based queries
- Add examples for viewing A2ATasks created by queries

dwmkerr committed:
refactor: consolidate A2A client creation logic

Removed duplicate A2A client creation code by exporting CreateA2AClient
from genai package. Both the A2ATaskReconciler and agent execution now
use the same client creation logic.

- Export CreateA2AClient for shared use
- Remove duplicate header resolution and HTTP client setup
- Remove duplicate a2aTaskRequestHandler (uses customA2ARequestHandler)
- Remove optional recorder checks (both callers always provide)
- Remove unused net/http import from controller
- Reduces code duplication by ~40 lines

dwmkerr committed:
chore: remove stale devspace reload comment

dwmkerr committed:
feat: add A2A metadata to Query status

Add status.a2a field to Query CRD containing contextId and taskId from
A2A protocol interactions. This prepares for moving A2A context tracking
from annotations to status to fix the reconciliation loop anti-pattern.

- Add A2AMetadata struct with contextId and taskId fields
- Add status.a2a field to QueryStatus
- Regenerate CRDs with make manifests
- Manually patch dist/chart CRD to preserve Helm templating

All tests pass. Next phase will update controller to write to status
instead of annotations.

dwmkerr committed:
fix: write A2A context ID to Query status instead of annotations

Moves A2A context ID from annotations to status.a2a.contextId to fix
reconciliation loop anti-pattern. Controller no longer calls r.Update()
on metadata during reconciliation, preventing unnecessary generation
increments and reconciliation triggers.

Changes:
- Write context ID to status.a2a.contextId instead of annotation
- Remove r.Update() call that triggered reconciliation loops
- Remove unused annotations import
- Update documentation to reflect status-based storage

dwmkerr committed:
feat: stream complete query status with A2A metadata in final chunk

Adds A2A metadata (contextId, taskId) to streaming responses by including
complete query status in the final chunk sent to clients.

Changes:
- Add QueryStatus field to StreamMetadata struct
- Update WrapChunkWithMetadata to accept optional query parameter
- Send final chunk with complete query status before stream completion
- Clients receive A2A metadata via ark.queryStatus.a2a in final chunk

This enables clients to access A2A context and task IDs without polling
the Query resource after streaming completes.

lucaromagnoli committed:
fix: stream errors properly in SSE format instead of closing stream

- Fix StreamingError to use OpenAI-compatible error format with error.message, error.type, error.code
- Stream error chunks when query execution fails in query_controller
- Stream error chunks when individual target (agent/team/model) execution fails
- Fix TypeScript stream handler to send errors as SSE events instead of JSON responses
- Ensure errors are sent before closing stream so OpenAI SDK and clients receive them

Fixes ARKQB-415: streaming now returns errors instead of silently closing the stream

lucaromagnoli committed:
refactor: address PR comments for streaming error handling

- Replace TODO with GitHub issue #415 reference for streaming improvement
- Extract error types to streaming.py: StreamingErrorDetail and StreamingErrorResponse
- Use typed error structure in proxy_streaming_response
- Remove parseErrorChunk fallback logic, validate StreamError structure directly
- Remove redundant try/catch in ark-cli createChatCompletionStream (errors propagate naturally)

Addresses PR feedback on error handling and type safety

lucaromagnoli committed:
Merge branch 'main' into fix/ARKQB-415-streaming-does-not-return-error

lucaromagnoli committed:
refactor: use expected error structure directly in proxy_streaming_response

- Remove fallback logic for error parsing
- Expect and validate error structure: {"error": {"message": "...", "type": "...", "code": "..."}}
- Only use default error format if structure is completely invalid
- Improves type safety and reduces ambiguity

lucaromagnoli committed:
refactor: use expected error structure directly without fallbacks

- Require both 'message' and 'type' fields in error structure
- Only use default error format if structure is completely invalid
- Remove intermediate variables, build error_data directly from response
- Addresses PR feedback to use expected structure and fail otherwise

lucaromagnoli committed:
fix: reduce nesting complexity in executeTarget error handling

Extract error handling logic into handleTargetExecutionError helper function
to resolve nestif linting error. Also fix gofmt formatting in a2a.go.

dwmkerr committed:
fix: move A2A metadata to status and fix A2ATask creation

- Move A2A context/task ID from annotations to Query status to fix reconciliation loop anti-pattern
- Add queryStatus field to streaming metadata for final chunk
- Update ark-cli TypeScript types to handle queryStatus in streaming
- Fix CLI crash when processing final chunk without choices array
- Fix A2ATask validation by setting required a2aServerRef in spec
- Update tests to read A2A metadata from status instead of annotations
- Update streaming documentation

This fixes the reconciliation loop issue where r.Update() on annotations triggered unnecessary reconciliation cycles.

lucaromagnoli committed:
Merge branch 'main' into fix/ARKQB-415-streaming-does-not-return-error

dwmkerr committed:
feat: add input-required and auth-required phases for A2ATask

- Add PhaseInputRequired and PhaseAuthRequired constants
- Update ConvertA2AStateToPhase to map to dedicated phases
- Update A2ATask CRD enum to include new phases
- Regenerate CRD manifests

dwmkerr committed:
refactor: improve A2A protocol conversion robustness

- Add PhaseUnknown constant for unrecognized A2A states
- Extract PartKindText/Data/File constants to avoid magic strings
- Update CRD enum to include unknown phase
- Return PhaseUnknown instead of PhaseRunning for unrecognized states

dwmkerr committed:
test: add comprehensive unit tests for A2A protocol conversion

- 100% coverage on state/phase conversion functions
- 100% coverage on part/artifact/history conversion
- 100% coverage on metadata conversion and merging
- Test all A2A protocol states including edge cases
- Test deduplication logic for artifacts and history
- Test timestamp parsing and phase transitions
- Overall genai package coverage increased from 8.3% to 12.7%

lucaromagnoli committed:
Merge branch 'main' into fix/ARKQB-415-streaming-does-not-return-error

lucaromagnoli committed:
refactor: address PR comments for streaming error handling

- Remove telemetry dependency from handleTargetExecutionError
- Extract StreamError helper function to avoid code duplication
- Use StreamError helper in query-level, target-level, and A2A error handling
- Move telemetry recording to caller of handleTargetExecutionError
- Remove redundant error streaming code

Addresses @havasik's review comments on PR #372

havasik commented on ark/internal/genai/a2a_protocol.go:
```diff
@@ -0,0 +1,291 @@
+	}
+
+	status.ProtocolState = string(task.Status.State)
+	status.SessionID = task.ContextID
```
might be worth just using contextid for this concept to avoid confusion

havasik commented on ark/internal/controller/a2atask_controller.go:
```diff
@@ -0,0 +1,188 @@
+
+	// Update Completed condition based on phase
+	switch newPhase {
+	case genai.PhasePending, genai.PhaseAssigned:
```
Conditions are meant to be orthogonal concepts - especially as we have a phase already no need for this many

dwmkerr replied:
there's only one condition - `Completed` - it is just set based on phase [false unless it is terminal]

havasik commented on ark/internal/genai/token_collector.go:
```diff
@@ -47,5 +50,18 @@ func (c *TokenUsageCollector) GetTokenSummary() TokenUsage {
 	c.mu.Unlock()
 }
+
+func (c *TokenUsageCollector) SetA2AContextID(contextID string) {
```
Probably shouldn't be TokenUsageCollector's responsibility to handle contextid set / get

dwmkerr replied:
Messaged you thinking it might be a follow up

havasik commented on ark/internal/genai/token_collector.go:
```diff
@@ -11,6 +11,9 @@ type TokenUsageCollector struct {
 	tokenUsages []TokenUsage
+	// A2A context ID collected deep in the query execution flow (similar to token usage).
+	// In time this structure will be generalized into a query state collector.
+	a2aContextID string
```
Doesn't look like the correct place for this

havasik commented on ark/api/v1alpha1/a2atask_types.go:
```diff
@@ -0,0 +1,228 @@
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// AssignedAgent references the agent that executed or is executing this task.
+	// +kubebuilder:validation:Optional
+	AssignedAgent *AgentRef `json:"assignedAgent,omitempty"`
```
Not sure if we need all these cross references (assigned agent, query ref etc.)

dwmkerr replied:
assigned agent i renamed but what do you think of `origin.agent` potentially (as optional?)

dwmkerr replied:
TTL is set from _completion_ time now BTW which means we could still accumulate potentially incorrectly, what do you think is better?

dwmkerr replied:
apols i set it to expire after completion, still not entirely confident we should do it at all

havasik commented on ark/internal/genai/a2a.go:
```diff
@@ -365,3 +379,81 @@ func resolveA2AHeaders(ctx context.Context, k8sClient client.Client, headers []a
 }
+
+// handleA2ATaskResponse handles A2A task responses by creating A2ATask resources
+func handleA2ATaskResponse(ctx context.Context, k8sClient client.Client, task *protocol.Task, agentName, namespace, queryName string, recorder record.EventRecorder, obj client.Object) error {
```
naming of this function is misleading

havasik commented on ark/api/v1alpha1/a2atask_types.go:
```diff
@@ -0,0 +1,228 @@
+	// After this duration, the resource may be automatically deleted.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default="720h"
+	TTL *metav1.Duration `json:"ttl,omitempty"`
```
Cleanup is missing after TTL. How do we know how long the server is keeping the task info?

dwmkerr replied:
There's not an a2a task version itself unfortunately, it's a server side decision but i've added our TTL here now still feels like this might evolve...

havasik commented on ark/api/v1alpha1/a2atask_types.go:
```diff
@@ -0,0 +1,228 @@
+type A2ATaskSpec struct {
+	// QueryRef references the Query that created this A2A task.
+	// +kubebuilder:validation:Required
+	QueryRef QueryRef `json:"queryRef"`
```
is it correct, binding a task to a single query? point of a2a tasks is to be able fulfill goals / intents, might be through multiple rounds of chat interaction

dwmkerr replied:
Honestly I'm not 100% - a task always originates (or should originate) with a query - although later we could also just pass a task ID and create from that I guess, which means that both the query and agent that originated might be optional - shall we just have something like `origin.query` and `origin.agent`? WDYT?

dwmkerr committed:
chore: work in progress

dwmkerr committed:
feat: merge main into feat/a2a-task-crd

- Add telemetry recording for token usage
- Add A2A context and task ID tracking to Query status
- Add ark metadata field to non-streaming responses
- Update printer columns for A2ATask
- Regenerate lock files
- Merge timeout logic for A2A client
- Merge StreamMetadata with both Annotations and QueryStatus
- Merge ChatUI with A2A context tracking and selector hooks

dwmkerr committed:
feat: add TTL cleanup and rename sessionId to contextId in A2ATask

- Implement TTL cleanup for A2ATask resources (deletes after TTL from creation time)
- Rename sessionId to contextId in A2ATaskStatus to match A2A protocol terminology
- Update both config/crd and dist/chart CRDs consistently
- Remove noisy INFO log messages from routine reconciliation events
- Add nolint comments for pre-existing cognitive complexity in tests
- Restore Kubernetes events (kept for kubectl get events debugging)

amanvr committed:
feat: Add sample workflow (#431)

- [ ] Follow the [contributor guide](./CONTRIBUTING.md)
- [ ] End-to-end tests pass
- [ ] Unit tests for essential parts of your code, e.g. APIs exposed via
SDKs or endpoints
- [ ] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person>
for <code, bug, docs, etc>"
- [ ] Linked issues #eg101

Co-authored-by: Aman Vivek Ramkumar <aman_vivek_ramkumar@external.mckinsey.com>

adriander committed:
feat: Migrate Marketplace to External Marketplace Repository (#428)

https://github.com/mckinsey/agents-at-scale-marketplace is public now.
all prerequisites are fulfilled

### Summary
Extracted ARK marketplace (Phoenix, Langfuse) into a dedicated
repository at https://github.com/mckinsey/agents-at-scale-marketplace.
This removes ~1,850 lines of code and allows the marketplace to evolve
independently.

### Key Changes

1. **Deleted `marketplace/` folder** - All services (Phoenix, Langfuse)
and their docs now live externally
2. **Updated devspace.yaml** - Changed from local paths to git
dependencies referencing the external repo
3. **Updated documentation** - Simplified docs to link to external
marketplace; removed build-time content copying
4. **Cleaned up dependencies** - Removed unused `fs-extra` and `rimraf`
from docs

### Usage
Reference marketplace services via DevSpace dependencies:
```yaml
dependencies:
  phoenix:
    git: https://github.com/mckinsey/agents-at-scale-marketplace
    tag: phoenix-v0.1.3
    subPath: services/phoenix
```

### Steps to verify PR
#### Phoenix
1. start with a clean minikube
2. uncomment in `devspace.yaml` phoenix (lines 31-34)
3. run `devspace dev`
4. monitor controller log: `kubectl logs -f deployment/ark-controller -n
ark-system | grep telemetry`
check for `2025-11-12T09:15:02Z INFO telemetry.config initializing OTEL
telemetry {"endpoint":
"http://phoenix-svc.phoenix.svc.cluster.local:6006", "service":
"ark-controller", "headers": ""}`
5. check on browser: http://phoenix-svc.phoenix.127.0.0.1.nip.io:8080/

#### Langfuse
1. start with a clean minikube
2. uncomment in `devspace.yaml` langfuse (lines 36-39)
3. run `devspace dev`
4. monitor controller log: `kubectl logs -f deployment/ark-controller -n
ark-system | grep telemetry`
check for `2025-11-12T08:36:20Z INFO telemetry.config initializing OTEL
telemetry {"endpoint":
"http://langfuse-web.telemetry.svc.cluster.local:3000/api/public/otel",
"service": "ark-controller" ...}
5. check on browser: http://langfuse.telemetry.127.0.0.1.nip.io:8080/

#### Docs
1. `cd docs && npm install && npm run dev`
2. check docs in http://localhost:3000/

## Checklist

- [x] Follow the [contributor guide](./CONTRIBUTING.md)
- [x] End-to-end tests pass
- [x] Unit tests for essential parts of your code, e.g. APIs exposed via
SDKs or endpoints
- [x] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person>
for <code, bug, docs, etc>"
- [ ] Linked issues #eg101

lucaromagnoli committed:
feat: add streaming error handling for A2A agents

- Add StreamError function and error types to streaming.go
- Stream errors when A2A execution fails
- Enables error messages to be streamed to clients when A2A tasks fail (e.g., countdown agent with negative number)

lucaromagnoli committed:
Merge remote-tracking branch 'origin/main' into feat/a2a-task-crd

dwmkerr committed:
chore: remove unused e2e and validation test files

- Remove test/e2e/ directory (447 lines) - unused locally, CI uses separate setup
- Remove test/validation-failures/ directory - redundant with chainsaw tests in tests/admission-failures/
- Remove e2e Makefile targets (setup-test-e2e, test-e2e, cleanup-test-e2e)
- Update documentation to remove e2e and validation test references

lucaromagnoli committed:
Merge branch 'fix/ARKQB-415-streaming-does-not-return-error' into feat/a2a-task-crd

# Conflicts:
#	ark/internal/genai/streaming.go
#	services/ark-api/ark-api/src/ark_api/api/v1/openai.py

lucaromagnoli committed:
Merge remote-tracking branch 'origin/feat/a2a-task-crd' into feat/a2a-task-crd

lucaromagnoli committed:
fix: improve error handling for streaming response failures

- Add proper error parsing with validation
- Use expected error structure from response
- Add fallback for invalid error responses
- Import StreamingErrorResponse type

lucaromagnoli committed:
Merge branch 'main' into feat/a2a-task-crd

havasik committed:
feat: query status spec fix (a2a) (#433)

## Checklist

- [ ] Follow the [contributor guide](./CONTRIBUTING.md)
- [ ] End-to-end tests pass
- [ ] Unit tests for essential parts of your code, e.g. APIs exposed via
SDKs or endpoints
- [ ] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person>
for <code, bug, docs, etc>"
- [ ] Linked issues #eg101

lucaromagnoli committed:
Merge branch 'main' into feat/a2a-task-crd

havasik committed:
fix: generated types check

havasik committed:
fix: tabs in docs

havasik committed:
fix: docs references

havasik committed:
fix: cleanup local changes

havasik committed:
fix: e2e tests matching new status spec of query

dwmkerr committed:
chore: remove redundant log line in A2A execution

dwmkerr committed:
chore: merge main into feat/a2a-task-crd

dwmkerr committed:
chore: remove stale code

dwmkerr committed:
chore: update mock-llm to version 0.1.22 in chainsaw tests

dwmkerr committed:
fix: wait for Completed condition instead of phase done in a2a-message-context test

dwmkerr committed:
chore: better docs / wait on 'completed' not 'phase'

dwmkerr committed:
fix: wait for Completed condition in remaining a2a tests

dwmkerr committed:
chore: update mock-llm to version 0.1.22 and simplify testing docs

dwmkerr committed:
docs: add verbose flag to troubleshooting and clean up debugging section

dwmkerr committed:
docs: update testing documentation structure

havasik committed:
feat: revert chainsaw retry

dwmkerr committed:
chore: update mock-llm to version 0.1.22 in chainsaw tests (#437)

## Summary
- Update all chainsaw test mock-llm installations to version 0.1.22

havasik committed:
fix: new a2a chainsaw assert condition

dwmkerr committed:
fix: replace phase: done with Completed condition in Query tests

Updated 15 chainsaw tests to check for Completed condition instead of phase: done. This prevents tests from timing out when queries fail, since the Completed condition is set on both success and failure, while phase: done is only set on success.

Tests fixed:
- agent-default-model
- agent-overrides
- agent-partial-tool
- agent-structured-output
- agent-tools
- filesystem-mcp
- jq-filter-test
- model-custom-headers
- model-properties
- queries
- query-label-selector
- query-parameters
- team-graph-selector
- team-graph
- team-selector

dwmkerr committed:
fix: correct mock-llm port from 8080 to 6556 in agent-default-model test

Mock-llm v0.1.22 runs on port 6556 by default, not 8080. The incorrect port
caused the readiness probe to fail continuously, preventing the test from
progressing past the pod readiness check.

dwmkerr committed:
refactor: split agent-default-model test into steps with targeted catch blocks

Break the monolithic test-with-default-model step into 4 logical steps:
- setup-mock-service: catches Pod failures
- create-default-model: catches Model availability issues
- create-agent-and-query: catches Agent readiness issues
- verify-query-completion: catches Query execution issues

This ensures failed assertions describe the actual failing resource rather than
attempting to describe downstream resources that don't exist yet.

havasik committed:
fix: segfault in team execution

dwmkerr committed:
fix: remove trailing newline from contextId in a2a-message-context test

Use 'echo -n' instead of 'echo' to prevent adding a newline character to the
contextId value. The newline was being embedded in the annotation value,
potentially breaking A2A context lookup.

dwmkerr committed:
fix: add fast timeouts for terminal state assertions in a2a-message-context

Add 1s timeout to assertions after Completed=True since the Query is in a
terminal state and won't change. Also add debug script to show actual response
content and contextId annotation when the test fails, making it easier to
diagnose A2A context propagation issues.

dwmkerr committed:
fix: use printf instead of echo -n to extract contextId

The 'echo -n' was being captured literally as '-n <contextId>' instead of
suppressing the newline. Use printf '%s' instead which reliably outputs
without a trailing newline and works across all shells.

dwmkerr committed:
debug: add response structure logging to a2a-message-context test

Show the full response structure to understand why the a2a.contextId field is
missing. This will help identify if the issue is in mock-llm not returning a
contextId or in the Ark controller not populating the field.

dwmkerr committed:
refactor: split a2a-message-context into steps with targeted catch blocks

Break the monolithic test into 4 logical steps:
- setup-mock-llm: catches A2AServer failures
- test-first-query: catches first Query failures
- extract-context-id: catches contextId extraction failures
- test-second-query-with-context: catches second Query failures

This ensures failed steps describe the actual failing resource rather than
attempting to describe resources that were never created due to earlier
failures.

dwmkerr commented on ark/internal/genai/a2a.go:
```diff
@@ -374,3 +407,82 @@ func resolveA2AHeaders(ctx context.Context, k8sClient client.Client, headers []a
+		a2aServerAddress = a2aServer.Status.LastResolvedAddress
+	}
+
+	labels := map[string]string{
```
no need for this, we're not yet using labels

dwmkerr commented on ark/internal/genai/a2a.go:
```diff
@@ -374,3 +407,82 @@ func resolveA2AHeaders(ctx context.Context, k8sClient client.Client, headers []a
+
+	annotations := make(map[string]string)
+	if a2aServerAddress != "" {
+		annotations["ark.mckinsey.com/a2a-server-address"] = a2aServerAddress
```
annotations should be consts

dwmkerr replied:
also no need to annotate with the address/server name, this is only for agents that need some back-linking, the a2atask points to the a2a originating agent anyway

dwmkerr commented on ark/api/v1alpha1/a2atask_types.go:
```diff
@@ -0,0 +1,226 @@
+	CompletionTime *metav1.Time `json:"completionTime,omitempty"`
+	// Error contains the error message if the task failed.
+	// +kubebuilder:validation:Optional
+	Error string `json:"error,omitempty"`
```
this error is not needed right, normally that'll be in the condition - I don't think we have error for query, pls check

dwmkerr replied:
Needed - it is part of the A2A task spec

dwmkerr committed:
chore: remove redundant log in A2A message execution

Remove superfluous log message 'Sending A2A message with context' as the
contextId is already logged earlier when extracted from the query annotation.

dwmkerr commented on ark/internal/genai/a2a.go:
```diff
@@ -374,3 +407,82 @@ func resolveA2AHeaders(ctx context.Context, k8sClient client.Client, headers []a
+					if queryName != "" {
+						return queryName
+					}
+					return "auto-generated-query" // Fallback for backwards compatibility
```
do not need backwards compat

dwmkerr committed:
chore: remove superfluous log in A2ATask TTL cleanup

Remove log statement after successful TTL-based task deletion. The deletion is
already observable through Kubernetes events and resource lifecycle.

dwmkerr commented on docs/content/reference/crds.mdx:
```diff
@@ -1,44 +1,192 @@
 ---
 title: Custom Resource Definitions (CRDs)
-description: Detailed specifications for ARK's custom resources
+description: Detailed specifications for Ark's custom resources
```
there should be no changes here except for one new item in the table

dwmkerr commented on tools/ark-cli/src/components/ChatUI.tsx:
```diff
@@ -606,11 +617,25 @@ const ChatUI: React.FC<ChatUIProps> = ({
       }
 
       // Send message and get response with abort signal
+      // Use ref to get the latest context ID, not the stale state from closure
```
no need for breadcrumbs

dwmkerr commented on tools/ark-cli/src/components/ChatUI.tsx:
```diff
@@ -606,11 +617,25 @@ const ChatUI: React.FC<ChatUIProps> = ({
 
       // Send message and get response with abort signal
+      // Use ref to get the latest context ID, not the stale state from closure
+      log('ChatUI: About to send message', {
```
eliminate lgo

dwmkerr committed:
chore: remove superfluous A2A contextId log statement

Remove redundant log 'A2A context ID from query'. The contextId is already
observable through events and is passed to executeA2AAgentMessage where it's
used to create the message with context.

dwmkerr committed:
fix: error if queryName missing when creating A2ATask

Validate that queryName is present before creating A2ATask resource. Without
a valid query name, the A2ATask cannot be properly linked to its originating
Query, breaking the relationship. Return error 'unable to determine A2A Task
originating query' instead of silently creating a broken reference.

dwmkerr committed:
chore: remove unused labels from A2ATask creation

Remove task-id and agent labels from A2ATask resources as they are not
referenced or used anywhere. The task ID is already in the resource name and
spec, and agent is in the spec.AgentRef.

dwmkerr committed:
chore: remove unused annotations from A2ATask creation

Remove a2a-server-address and a2a-server-name annotations from A2ATask
resources as they are not referenced or used anywhere. The A2A server
information is already available in spec.A2AServerRef.

lucaromagnoli committed:
fix: add timeouts to chainsaw test assertions

- Add explicit timeouts to agent-default-model test assertions
- Add timeouts to filesystem-mcp-workspace-isolation query phase assertions
- Increase query phase timeout to 240s to allow LLM calls to complete
- Fixes flaky test failures in CI

lucaromagnoli committed:
fix: address PR review comments

- Remove backwards compatibility fallback in a2a.go (queryName should always be provided)
- Remove log statements from ChatUI.tsx
- Restore crds.mdx to match main, only adding A2ATask entry to table

dwmkerr commented on ark/internal/genai/a2a_execution.go:
```diff
@@ -80,8 +80,13 @@ func (e *A2AExecutionEngine) Execute(ctx context.Context, agentName, namespace s
 		content = userInput.OfUser.Content.OfString.Value
 	}
 
+	if contextID != "" {
```
not needed

dwmkerr committed:
docs: update chainsaw test selector examples

Add selector documentation for skipping tests that require images and improve
formatting of command examples with line continuations.

dwmkerr committed:
chore: merge chore/test-e2e into feat/a2a-task-crd

dwmkerr committed:
chore: remove debug logging and clarify streaming metadata logic

- Fix chainsaw test variable scoping for A2A context ID extraction
- Add clarifying comment about streaming metadata field override behavior
- Remove unnecessary debug log statements from ark-api and ark-cli
- Remove unused debug.ts and test-metadata.js files

dwmkerr committed:
chore: remove unnecessary comment about refs

dwmkerr committed:
refactor: simplify streaming metadata by including full Query object

Replace queryStatus field with completedQuery in streaming metadata to provide
complete query information including metadata, spec, and status. This eliminates
manual field copying and gives clients access to all query data.

Changes:
- Replace StreamMetadata.QueryStatus with CompletedQuery (full Query object)
- Remove unused Annotations field from StreamMetadata
- Simplify WrapChunkWithMetadata - just set query instead of copying fields
- Update CLI to access completedQuery.status.a2a.contextId
- Update documentation with new field structure and examples

dwmkerr committed:
docs: remove duplicate OpenAI-Compatible APIs section

Consolidated two duplicate sections into single comprehensive section
with detailed examples and response formats.

dwmkerr committed:
chore: remove superfluous logging in openai.py

Removed debug logging for queryAnnotations parsing and application

dwmkerr committed:
chore: remove extra timeouts from agent-default-model test

Removed explicit timeouts to match main branch pattern

dwmkerr committed:
chore: remove extra timeouts from filesystem-mcp test

Removed explicit timeouts added for query assertions to match main branch pattern

dwmkerr committed:
fix: extract A2A context ID from response object

A2A context ID is stored in status.responses[0].a2a.contextId, not
status.a2a.contextId. Updated ChatUI to extract from correct path.

Chat TUI always queries a single target, so contextId is in responses[0].

dwmkerr committed:
fix: update streaming tests for CompletedQuery field

Updated streaming_test.go to test CompletedQuery instead of removed
Annotations field. Tests now verify full Query object is passed.

dwmkerr commented:
3x reviewers and testers merging in now
